pytential/symbolic/execution.py

```diff
@@ -574,12 +608,40 @@ def auto_source(self):
     def auto_target(self):
         return self.auto_where[1]
 
-    def _get_lpot_discretization(self, lpot, dofdesc):
-        if dofdesc.discr_stage == sym.QBX_SOURCE_STAGE2:
-            return lpot.stage2_density_discr
-        if dofdesc.discr_stage == sym.QBX_SOURCE_QUAD_STAGE2:
-            return lpot.quad_stage2_density_discr
-        return lpot.density_discr
+    @property
+    @memoize_method
+    def ambient_dim(self):
+        from pytools import single_valued
+        ambient_dim = [p.ambient_dim for p in six.itervalues(self.places)]
+        return single_valued(ambient_dim)
+
+    def _get_qbx_discretization(self, dofdesc):
+        lpot_source = self.get_geometry(dofdesc)
+        if lpot_source._disable_refinement:
+            return lpot_source.density_discr
+
+        if dofdesc.discr_stage is None:
+            dofdesc = dofdesc.to_stage1()
+
+        # NOTE: need to keep cache name in sync with `_refine_for_global_qbx`
+        cache = self.get_cache('refined_qbx_discrs')
```

Comment:
Make these cache names variables that are stored somewhere, instead of strings that are equal "by accident".

Reply:
Yes! What's the preferred way here? I can add some private global variables/objects somewhere and use those as keys consistently. Would that work?